tests/sim_trace_viz.py

```diff
@@ -0,0 +1,164 @@
+    "BecomeLeader": "BLea",
+    "become_follower": "BFol",
+    "AdvanceCommitIndex": "Cmt",
+    "bootstrap": "Boot",
+    None: "",
+    "SignCommittableMessages": "Rpl",
+    "SIMTimeout": "Tmou",
+    "RcvUpdateTerm": "UTrm",
+    "SIMCheckQuorum": "ChkQ",
+}
+
+TAG = {"Y": ":white_check_mark:", "N": ":x:", " ": "  ", "S": ":pencil:"}
+
+
+def digits(value):
+    return len(str(value))
+
+
+def diffed_key(old, new, key, suffix, size):
+    if old is None or old[key] == new[key]:
+        return f"{new[key]:>{size}}{suffix}"
+    color = "bright_white on red"
+    return f"[{color}]{new[key]:>{size}}{suffix}[/{color}]"
+
+
+def render_state(state, func, old_state, tag, cfg):
+    if state is None:
+        return " "
+    ls = LEADERSHIP_STATUS[state["leadership_state"]]
+    nid = state["node_id"]
+    v = diffed_key(old_state, state, "current_view", "", cfg.view)
+    i = diffed_key(old_state, state, "last_idx", "", cfg.index)
+    c = diffed_key(old_state, state, "commit_idx", "", cfg.commit)
+    f = FUNCTIONS[func]
+    opc = "bold bright_white on red" if func else "normal"
+    return f"[{opc}]{nid:>{cfg.nodes}}{ls}{f:<4} [/{opc}]{TAG[tag]} {v}.{i} {c}"
+
+
+class DigitsCfg:
+    nodes = 0
+    view = 0
+    index = 0
+    commit = 0
+    ts = 0
+
+
+def extract_node_state(global_state, node_id):
+    return {
+        "node_id": node_id,
+        "leadership_state": global_state["leadershipState"][node_id],
+        "current_view": global_state["currentTerm"][node_id],
+        "last_idx": len(global_state["log"][node_id]),
+        "commit_idx": global_state["commitIndex"][node_id],
+    }
+
+
+LOG = {
+    "Reconfiguration": ":recycle:",
+    "Signature": ":pencil:",
+    "Entry": ":page_facing_up:",
+}
+
+
+def render_log(log, dcfg):
+    term = 2
+    chars = []
+    for entry in log:
+        if entry["term"] != term:
+            term = entry["term"]
+            chars.append(f" {term:>{dcfg.view}}")
+        chars.append(LOG[entry["contentType"]])
+    return "".join(chars)
+
+
+def table(entries):
+    nodes = []
+    max_view = 0
+    max_index = 0
+    max_commit = 0
+    max_ts = 0
+    rows = []
+    for pre, action, post in entries:
+        node_id = action["context"]["i"]
+        if node_id not in nodes:
+            nodes.append(node_id)
+        max_view = max(max_view, post[1]["currentTerm"][node_id])
+        max_index = max(max_index, len(post[1]["log"][node_id]))
+        max_commit = max(max_commit, post[1]["commitIndex"][node_id])
+        max_ts += 1
+    dcfg = DigitsCfg()
+    dcfg.nodes = len(max(nodes, key=len))
+    dcfg.view = digits(max_view)
+    dcfg.index = digits(max_index)
+    dcfg.commit = digits(max_commit)
+    dcfg.ts = digits(max_ts)
+    ts = 0
+    for pre, action, post in entries:
+        ts += 1
+        if action["name"] in {"SIMTimeout", "SIMCheckQuorum", "UpdateTerm"}:
+            continue
+        node_id = action["context"]["i"]
+
+        def last_recvd():
```

Comment:
@lemmy as discussed, it would be great if there was a way to easily add the result of some expressions to the trace, in this case the last received message.

Reply:
Do we need support for action-level or just state-level formulas.

Reply:
Action-level seems preferable here, since the goal is to report the message(s) being consumed in the action. The way the script does it is by diffing the messages queues from the pre/post states, so depending on what the formula has access to at the state level, that may be fine too.

Reply:
The feature should be ready in a few days.

Reply:
Thank you! I think that will make the traces much easier to read, both in TLA directly (where looking up the details of the messages consumed in the action is not straightforward now), and for tools like this one.

Reply:
@lemmy yes but the output of ALIAS doesn't make it to the JSON trace produced by `-dumpTrace json file.json` as far as I can tell from a quick test with the DebugAlias on the SIM.

Reply:
Yes, that the part I'm working on. My confusion is around
> ...in TLA directly (where looking up the details of the messages consumed in the action is not straightforward now

Reply:
Sorry, I wasn't clear, that referred to adding an alias that would diff the messages, which I don't think we have yet. But that's something I can do already, you are right.

Reply:
Done: https://github.com/tlaplus/tlaplus/pull/865